beps/bep_0007.rst

```diff
@@ -23,32 +24,31 @@ be interpreted as described in IETF `RFC 2119`_.
 
 .. _`RFC 2119`: http://tools.ietf.org/html/rfc2119
 
-Announce Parameter
-==================
+Announcing
+==========
+
+For each local IP address:
 
-The client MAY add an ``&ipv6=`` parameter to the HTTP GET request it sends
-to the tracker. The value is either an IPv6 endpoint (address and port) or
-just an IPv6 address. In the case where only an address is supplied, the IPv6
-port is assumed to be the same as specified by the ``&port=`` parameter.
+	1. the client intends to receive incoming peer connections over (listens on)
+	2. the client intends to publish to the tracker
+	3. that can route packets to the tracker
 
-The tracker SHOULD perform a NAT check on the IPv6 endpoint.
+an announce to the tracker SHOULD be made. Each announce should use the
+corresponding local IP address as the source address of the connection.
```

Comment:
What are you mean by *local* IP address? Shouldn't here be *external* IP address?

If I listen on 192.168.0.2:50000 and I available from internet by 100.100.100.100:50001, which address/port should I announce according to that instruction?

Reply:
Well, this is about multi-homing. Which is mostly a thing on IPv6 (which this BEP is about) or datacenter servers which may have multiple public v4 addresses.

Multi v4 behind NAT is a complex edge-case that's out of scope for this BEP imo.

Reply:
If it is only about ipv6 then shouldn't it be purely stated?

Withal, as I understand, this sentences not only ipv6 but about ipv4 too. Am I mistaken? I think that in case it is about both types of addresses than 'external' would be more vital because it will be the same as local for the ipv6 addresses and it will be correct about ipv4 addresses.

Reply:
> What do you mean by local IP address? Shouldn't here be external IP address?

No. You don't necessarily know what your external address is, but you *do* know your local IP addresses, that you can listen on and that are reachable from the internet.

> If it is only about ipv6 then shouldn't it be purely stated?

It applies to both IPv4 and IPv6. The latter is probably more common in practice.

> I think that in case it is about both types of addresses than 'external' would be more vital because it will be the same as local for the ipv6 addresses and it will be correct about ipv4 addresses. What do you think?

You can't bind a listen socket (normally) to your "external" address. As long as "external" means one that's addressable on the internet. The tracker will know what your external address is based on the source address in the IP header.

In your mind, what would the implication be to change the wording to say "external" instead?

Reply:
Okay, I think I got it.

I wasn't attentive enough and forgot about the fact that IP address isn't sending now. And here you are talking about source IP and not the sent to tracker.

Thank you.

Reply:
**I think the whole logic is wrong and with DualStack being real, we have to come up with an alternate solution, as painful as it might be!**

In the legacy-IP (i.e. IPv4-only) world we had (except multi-homed) only one outward connection. So it was fine for the tracker to assume that this is the peer's IP. But in DualStack land the tracker will receive the connection via IPv6 or IPv4, depending on what's available and what's preferred (i.e. tracker IPv4/IPv6?, system preferences, routing difficulties). However, except for rare cases the client will not know its own external addresses, because of:

1. NAT and CGNAT
  --> if the client connects to the tracker via IPv6, how should the tracker know the public IPv4 of the client? And if, the client most likely will be behind NAT (home server behind router) or CGNAT (incoming connections impossible [except for some very rare cases but those are neglectable]). So there are only two options to pass the public IP in the GET request:
_a) specify IPv4 manually (tedious, if on dynamic connection, be it fixed or VPN)_
_b) use external service to determine your public IPv4 (e.g. https://ifconfig.co/)_

2. GUA, temporary, ULA, link-local, NAT
  --> if the client connects to the tracker via IPv4, which IPv6 should it chose to announce in the GET request? There's probably at least the fixed GUA and the temporary GUA (i.e. privacy extensions), but there might also be ULA, link-local addresses, and yes, sometimes there even might be NAT (mostly seen with IPv6-capable VPN). While at least the fixed GUA should work, it might not be preferred by the user due to privacy concerns. In Linux you can prefer the temporary address in userspace, not sure how that works in Windows. ULA might be interesting only for local testing, but might be necessary with NAT (in IPv6 environments, NAT is a simple prefix translation, not port translation, so it's more comparable to static- or 1:1-NAT in the IPv4 world).
_a) You could allow the user to define to prefer fixed addresses
b) You could allow the user to define to prefer temporary addresses
c) You could also allow the user to define the address to use, but this won't solve it in the rare case of NAT_

**But how do we solve this?** I strongly believe any logic to automate this process is complicated and will eventually be flawed. The only reasonable approach I see is to send two separate GET requests to the tracker, one from each address family. The tracker will then see the respective IPs that the requests are coming from and should be using these to announce the peer to other peers. Consequently, I am suggesting to drop the &ipv4, &ipv6, and &ip parameters.

**What to do with DualStack hosts on SingleStack trackers?** Well, now we have another problem: If we assume we have some DualStack hosts that aren't reachable over IPv4 due to CGNAT, but which are fine serving requests over IPv6, we want them to participate. Unfortunately, our single-stack tracker is IPv4-only, so our client somehow has to tell the tracker its IPv6. If we want this, I think we need to have either (like mentioned before) some 3rd-party service, or allow the user to configure this manually in the client. The other way round (DualStack hosts on IPv6-only tracker) should be rather unproblematic (because single-stack hosts most likely would be IPv4-only and wouldn't be able to connect to the tracker anyways).

Well, probably I missed some very important point or have a serious flaw in my logic as well. Hope we can find the best solution by discussion.

Reply:
I think you're arguing in favour of this change of BEP 7, is that correct?
Depending a bit on what you mean by: "I strongly believe any logic to automate this process is complicated and will eventually be flawed.".

The solution documented by this patch to BEP 7 (and what libtorrent is doing in practice) *is* automating "this", as long as "this" refers to "getting both your IPv4 and IPv6 address in the tracker peer list".

It *is* pretty complicated though. I tried to document how I'm changing libtorrent to do this for the next release [here](https://github.com/arvidn/libtorrent/pull/4267) if anyone is interested.

Reply:
Probably, but then I didn't really understand how the client is supposed to bind to a specific address, if it doesn't know its own addresses. At least the client won't know its address if it's behind NAT. That applies both for IPv4 and IPv6, though NAT on IPv6 is rather rare, and I believe also multi-homed hosts. This seems to be complicated by the fact that only one &ipv4 and one &ipv6 is allowed? So in those cases we either need to:

- make connection to tracker using corresponding address family,
- use some external service, or
- specify IPs manually

For IPv6 we also need to decide which address to use, if privacy extensions are enabled. Windows will prefer the temporary PE address. While it's easy on Linux to chose which address to bind to on outgoing connections, it's not implemented in Windows. The only(!) solution for this I have found on the web is documented here:
https://stackoverflow.com/questions/35442119/find-temporary-ipv6-address-using-getadaptersaddresses?rq=1

But yes, I also read that trackers handle this different as well. Some will drop any previous addresses from a peer if a new announce is send with the same &key. So I believe this is something that needs both support from clients and trackers. I don't see at the moment what would be the best practice with current implementations in mind. Best fallback would probably be to try to send the request using both &ipv4 and &ipv6 in requests from both address families and just assume that one IPv4 and one IPv6 is good enough in most cases.

And I think we need to solve all this and make it work correctly on DualStack, and only then can we think about scenarios where one might want to suppress IPv6 due to leakage concerns etc.

Reply:
> Probably, but then I didn't really understand how the client is supposed to bind to a specific address, if it doesn't know its own addresses.

Clients don't necessarily know their *external* addresses. But you don't bind sockets to external addresses anyway. You bind sockets to *your* local addresses, which are known.

> At least the client won't know its address if it's behind NAT. That applies both for IPv4 and IPv6, though NAT on IPv6 is rather rare, and I believe also multi-homed hosts. This seems to be complicated by the fact that only one &ipv4 and one &ipv6 is allowed?

This patch (this pull request you're commenting on) removes the `&ipv4=` and `&ipv6=` query parameters. Is that comment referring to the current wording of BEP7?

> And I think we need to solve all this and make it work correctly on DualStack, and only then can we think about scenarios where one might want to suppress IPv6 due to leakage concerns etc.

You suggest that it isn't already solved. I'm suggesting that this patch solves "the problem". Could you please highlight, specifically, what problem is *not* solved by this updated wording of BEP7 that is proposed by this patch?

Reply:
You are right and reading it once again now made it much clearer. I was probably already thinking too much implementation-wise than simply protocol-wise. So yes, the way that BEP7 is written now seems good. I didn't see any flaws and approve the change. Especially the "SHOULD NOT" in "If the tracker does not resolve to a an address family compatible with a source address, the client SHOULD NOT announce that address to the tracker." is perfect in my opinion, as it should be possible in general, but tracker and client should really think about what they are doing when they offer this option.